### PR TITLE
Max parallel upperbound

### PR DESCRIPTION
**Pull request by disparter**
[DO NOT MERGE IT - SCRATCH PHASE]

Summary:
[Enforce maximum parallelization factor for large stages (based on N number of instances)]
  - For single side car
  - For multiple sidecars
  - For all configured stages
   
   
Test Plan:
Create unit tests to clear the math behind parallelization counts
Deploy N numbers above constraints and see if holds under all parallelization rules

### CODE REVIEW

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java [hunk 1]
```diff
@@ -227,7 +229,7 @@ boolean canDeploy(EnvironBean envBean, String host, AgentBean agentBean) throws
         String envId = envBean.getEnv_id();
         AgentCountBean agentCountBean = agentCountDAO.get(envId);
         long totalNonFirstDeployAgents = (isAgentCountValid(envId, agentCountBean) == true) ? agentCountBean.getExisting_count() : agentDAO.countNonFirstDeployingAgent(envId);
-        long parallelThreshold = getFinalMaxParallelCount(envBean, totalNonFirstDeployAgents);
+        long parallelThreshold = extracted(envBean, totalNonFirstDeployAgents);
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java:232]: what's extracted ?

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java [hunk 2]
```diff
@@ -822,4 +825,20 @@ public final static int getFinalMaxParallelCount(EnvironBean environBean, long t
         }
         return ret;
     }
+
+    public final static int getFinalMaxParallelCount(EnvironBean envBean, long totalNonFirstDeployAgents, int maxParallelThreshold) throws Exception{
+        return Math.min(maxParallelThreshold, getFinalMaxParallelCount(envBean, totalNonFirstDeployAgents));
+    }
+
+    public final static int getFinalMaxParallelCount(DeployConstraintBean deployConstraintBean, EnvironBean envBean, int totalHosts) throws Exception{
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java:833]: where is this used?

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java (hunk 1 above)
Comment (disparter) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java:232]: This PR is already closed after our discussion.
But I extracted the method to be able to unit testing the calculation without the need of complex testing for this.
The name was unhappy.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java (hunk 2 above)
Comment (disparter) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java:833]: This is the extracted method name from the above question. The refactor was misplaced.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java (hunk 1 above)
Comment (ankilosaurus) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java:232]: yeah, I realized that it's closed. Please ignore comments :)

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java (hunk 2 above)
Comment (ankilosaurus) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java:833]: plz ignore

**Pull request closed without merging**